### ISSUE TITLE
git+ssh dependencies

### ISSUE THREAD

**Issue by eqyiel** [Profpatsch/yarn2nix#3]
`yarn2nix` seems to fail for lockfiles including (private) git+ssh dependencies: 

```
$ yarn2nix
yarn2nix: Map.!: given key is not an element in the map
CallStack (from HasCallStack):
  error, called at libraries/containers/Data/Map/Internal.hs:603:17 in containers-0.5.10.2:Data.Map.Internal
```

Is this expected?  If not, I'm very interested in helping to debug this (I don't know much about Haskell though).

**Comment by eqyiel** [comment]
Hm, it looks like it's caused by dependencies with part of the sha1 hash at the end of the URL:

```
"mango-components@git+ssh://git@github.com:mango-chutney/mango-components.git#f617aa4":
  version "1.0.3"
  resolved "git+ssh://git@github.com:mango-chutney/mango-components.git#f617aa4bd97d2da500afeb135f7c62ae8b42ace7"
  dependencies:
    babel-polyfill "^6.26.0"
    classnames "^2.2.5"
    polished "1.8.0"
    popper.js "^1.12.6"
    react "^16.0.0"
    react-dom "^16.0.0"
    react-popper "^0.7.4"
    react-portal "4.0.0-beta.2"
    react-redux "^5.0.6"
    redux "^3.7.2"
    redux-form "^7.1.1"
    tristicons "git+ssh://git@github.com:mango-chutney/tristicons.git#a8bfdd8"
```

This is what appears in the lockfile after something like `yarn add "git+ssh://git@github.com/mango-chutney/mango-components.git#f617aa4"`.

**Comment by Profpatsch** [comment]
Thanks for the example, I will take a look.

I notice I should put a few more call stack annotations into my multi-keyed map library; it looks like it is still buggy.

**Comment by Profpatsch** [comment]
So it turns out parsing the identifiers is ambiguous. I [submitted an issue upstream](https://github.com/yarnpkg/yarn/issues/5333).

@eqyiel What version of `yarn` generated this `yarn.lock` file?

**Comment by eqyiel** [comment]
@Profpatsch thanks for looking into it!

The yarn version is `v1.3.2`.

I put together a minimal example:

https://github.com/eqyiel/yarn2nix-example-a
https://github.com/eqyiel/yarn2nix-example-b
https://github.com/eqyiel/yarn2nix-example-c

`yarn2nix-example-a` depends on `yarn2nix-example-b` which depends on `yarn2nix-example-c`.

Note that `yarn2nix` succeeds if run in `yarn2nix-example-b` and `yarn2nix-example-c` but not in `yarn2nix-example-a`.

It's set up so that you can try it out using the same `nixpkgs` as me after running nix-shell in those repos (if you don't use [`direnv`](https://github.com/direnv/direnv), just source the `.envrc` file).

**Comment by Profpatsch** [comment]
Okay, so after researching for a few hours, I found out the reason for this confusion is [npm scopes](https://docs.npmjs.com/misc/scope) which allow for one additional `@` at the beginning of a package name (no comment, really).
Yarn implements that with a [special case in their parsing of package identifier strings](https://github.com/yarnpkg/yarn/blob/2d454b552d447a0f79a04e4e451e926e1c0a29e7/src/util/normalize-pattern.js#L17). We should do the same, meaning the [`pkgKey`](https://github.com/Profpatsch/yarn-lock/blob/98728b4335e269ad2bfc63ca9bc662c5c0a4eede/src/Yarn/Lock/Parse.hs#L107) function should be adjusted accordingly, so it allows `@` in versions as well and only at the beginning of package names.

The confusing error stems from the parser splitting the package name on the wrong `@`, I [added a test](https://github.com/Profpatsch/yarn-lock/commit/98728b4335e269ad2bfc63ca9bc662c5c0a4eede) that shows this:

```
    complexKey:              FAIL
      complexKey name
      expected: "mango-components"
       but got: "mango-components@git+ssh://git"
```

That means the resulting [`MuiltiKeyedMap`](https://github.com/Profpatsch/yarn-lock/blob/master/src/Data/MultiKeyedMap.hs) doesn’t have a key with index name `"mango-components"` and fails at the [dereference in the `decycle` algorithm](https://github.com/Profpatsch/yarn-lock/blob/98728b4335e269ad2bfc63ca9bc662c5c0a4eede/src/Yarn/Lock/Helpers.hs#L45) because the [invariant](https://github.com/Profpatsch/yarn-lock/blob/98728b4335e269ad2bfc63ca9bc662c5c0a4eede/src/Yarn/Lock/Helpers.hs#L31) does not hold.

Fixing the parser should be fairly trivial, if you’d like to take a shot.

Additionally, the support for npm scopes needs to be implemented, see #4.

**Comment by Profpatsch** [comment]
Solved by https://github.com/Profpatsch/yarn-lock/commit/5460e22abe07f6f688757bf63b9b33bce710c063. Also, scopes are implemented now. :balloon:
